deploy/README.md

```diff
@@ -1129,3 +1129,246 @@ Ensure that no keys duplicated across `gramex.yaml` files and restart gramex.
   or refer to a file / folder that cannot be accessed, etc.
 - `Gramex 1.xx.x is available`: A newer version of Gramex is available. Upgrade
+
+## Good security practices
+
+A curated list of security practices for real-life application deployments is presented here.
+
+### Session timeout
+
+Applications may need to timeout user sessions periodically. We can configure this as below:
+
+```yaml
+# gramex.yaml
+application/login:
+  pattern: /$YAMLURL/login
+  handler: DBAuth
+  kwargs:
+    session_expiry: 0.0207           # session expires after 30 mins
+    session_inactive: 0.0207         # session expires after 30 mins of inactivity
+```
+
+### Disable directory listing
+
+Directory listing isn't a recommended practice as it reveals exact file names. One can disable directory listing at a root level as below:
+
+```yaml
+# gramex.yaml
+handlers:
+  FileHandler:
+    ignore:                             # accepts a list of files to be ignored
+      - '*.yaml'                        # YAML defines routes, user credentials
+      - '*.git*'                        # git files
+      - '*.json'
+      - '*Dockerfile'
+      - '*nohup.out'                    # logs
+      - '*.git/*'
+      - '*ui/*'
+```
+
+### Protect all pages with authentication
+
+Templates are routinely used to render or refresh views on the interface. It's important to disable access without the correct authentication credentials and access controls.
+
+```yaml
+# gramex.yaml
+url:
+  templates-home:
+    pattern: /$YAMLURL/templates/(.*)
+    handler: FileHandler
+    priority: 100
+    kwargs:
+      path: $YAMLPATH/templates/
+      auth:
+        login_url: /$YAMLURL/login
+      headers: $FILEHANDLER
+```
+
+This is also true for any other files (static assets, data files, queries, remote function calls etc.).
+
+### Custom error messages
+
+Error messages can be customized based on their type. To do that, define a route then write a `Python` function that accepts error status code and handler as arguments.
+
+```yaml
+# gramex.yaml
+BaseHandler:
+  error:
+    500: &ERROR
+      function: app_script.error_fn(status_code, kwargs, handler)
+      header:
+        Cache-Control: no-cache
+        Content-Type: text/html
+    400: *ERROR
+    401: *ERROR
+    403: *ERROR
+    404: *ERROR
+```
+
+```py
+"""
+This is `app_script.py`.
+"""
+import os
+import gramex.cache
+
+
+def error_fn(status, kwargs, handler):
+    """Load the error pages as required.
+    Args:
+      kwargs (dict): keyword arguments
+      status (int): status code
+      handler (request object): gramex handler object
+    Returns:
+      (template): dynamic error page
+    """
+
+    error_path = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), 'errorpage.html')
+    tmpl = gramex.cache.open(error_path, 'template')
+    handler.set_status(status)
+    return tmpl.generate(kwargs=kwargs, status=status, handler=handler)
```

Comment:
Does this work or do we have to do a `decode('utf8')` on the generated template?

Reply:
TBH, didn't test it. will test and update in another commit (in a few days).

Reply:
this is now resolved.